### ISSUE TITLE
pushNamedAndRemoveUntil does not resolve future unless the new route gets popped

### ISSUE THREAD

**Issue by michelengelen** [flutter/flutter#44896]
## Steps to Reproduce
1. Build a navigation using the build in navigator with at least 2 screens
2. navigate around a bit
3. call `pushNamedAndRemoveUntil` with a false predicate to clear the navigation stack:
```dart
    final GlobalKey navigatorKey = GlobalKey<NavigatorState>();
    final NavigatorState nav = navigatorKey.currentState;
    nav.pushNamedAndRemoveUntil(Routes.home, (Route<dynamic> route) => false).then((Object _) => print('#### FUTURE RESOLVED'));
```
4. The Future returned by the `pushNamedAndRemoveUntil` method is not resolved
5. It will resolve ones you pop the new route of the stack

I do have an implementation of this in a redux middleware and the problem is, that I cannot determine when the transition of the navigation is finished before altering the store. I did it like this:

```dart
Middleware<RootState> _signOutUser(GlobalKey navigatorKey) {
  return (Store<RootState> store, dynamic dynamicAction, NextDispatcher next) async {
    final SignOut action = dynamicAction;
    final NavigatorState nav = navigatorKey.currentState;

    next(action);

    store.dispatch(AppIsLoading());

    Future<dynamic>(() async {
      try {
        await auth.signOut();
      } on PlatformException catch (error) {
        return Future<void>.error(error);
      }
    }).then<void>((dynamic _) {
      nav.pushNamedAndRemoveUntil(Routes.home, (Route<dynamic> route) => false);
      store.dispatch(UnsetUser());
      store.dispatch(AppIsLoaded());
      action.completer.complete();
    }).catchError((Object response) {
      nav.maybePop(nav.context);
      action.completer.completeError(response);
      store.dispatch(AppIsLoaded());
    });
  };
}
```

Ideally, since `pushNamedAndRemoveUntil` does return a future according to the docs, I would have done it like this to check for the operation to end, before unsetting the user.

```dart
Middleware<RootState> _signOutUser(GlobalKey navigatorKey) {
  return (Store<RootState> store, dynamic dynamicAction, NextDispatcher next) async {
    final SignOut action = dynamicAction;
    final NavigatorState nav = navigatorKey.currentState;

    next(action);

    store.dispatch(AppIsLoading());

    Future<dynamic>(() async {
      try {
        await auth.signOut();
      } on PlatformException catch (error) {
        return Future<void>.error(error);
      }
    }).then<void>((dynamic _) {
      return nav.pushNamedAndRemoveUntil(Routes.home, (Route<dynamic> route) => false);
    }).then<void>((dynamic _) {
      store.dispatch(UnsetUser());
      store.dispatch(AppIsLoaded());
      action.completer.complete();
    }).catchError((Object response) {
      nav.maybePop(nav.context);
      action.completer.completeError(response);
      store.dispatch(AppIsLoaded());
    });
  };
}
```

The underlying problem is, that the User is getting unset before the transition ends. This does pose a problem when navigating away from a site where data from this is needed to render the view, such as `displayName` or `photoUrl`.

Maybe there is a better way I just do not know about and I am very happy if there is and someone could point me to the right direction, but currently I do not see anything that could help me with it.

Now I did some digging in the `Navigator` and `Route` classes and found, that there is something wrong with the method and a possible solution to it. In `NavigatorState`s `pushNamedAndRemoveUntil` method (and several others) the value returned is `newRoute.popped`, which is a completer, that gets resolved once the newly pushed route gets popped of the stack.

While there is actually a method on the `Route` class called `didPush` that returns a resolved `TickerFuture` once the transition animation is done.

For now I solved this temporarily (for my project) with replacing
```dart
return newRoute.popped;
```
with 
```
return newRoute.didPush() as Future;
```

I had to cast `Future` onto it, because `newRoute.didPush` returns a `TickerFuture` and the Editor will yell at me if I do not do it that way! :D

Any help here is highly appreciated.

**Target Platform:** iOS
**Target OS version/browser:** 10.1
**Devices:** iPhone5S or higher

## Logs
`run` logs not needed, since it does not run into an exception

```
Analyzing sps...
No issues found! (ran in 3.5s)
```

```
[✓] Flutter (Channel stable, v1.9.1+hotfix.5, on Mac OS X 10.15 19A602, locale en-GB)
    • Flutter version 1.9.1+hotfix.5 at /Users/michel.engelen/Desktop/__flutter/_sdk
    • Framework revision 1aedbb1835 (4 weeks ago), 2019-10-17 08:37:27 -0700
    • Engine revision b863200c37
    • Dart version 2.5.0


[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/michel.engelen/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 11.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.1, Build version 11A1027
    • CocoaPods version 1.8.3

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 40.2.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (1 available)
    • iPhone 11 Pro Max (simulator) • 6064309D-727C-4BC4-A26A-EC07AEF14F0E • ios • com.apple.CoreSimulator.SimRuntime.iOS-13-1
      (simulator)

! Doctor found issues in 1 category.
```

**Comment by janmoppel** [comment]
Hello @michelengelen !

Thank you for posting an issue.
Your question might be a good fit for the [StackOverflow](https://stackoverflow.com/tags/flutter) as well. Flutter has a very active and helpful community there, so feel free to post it there too.

Also, see other support options here: https://flutter.dev/community

**Comment by michelengelen** [comment]
Hello @janmoppel.
I already did that and sadly there was no help coming from that side. So I wanted to make sure that the actual devs can have a look, since I think that the return value is not what the developers would expect form a method like this.

**Comment by michelengelen** [comment]
So, there is no update from dev-side?
I suggest this is an issue that is not coming up very often, but should be handled anyways.

If I am mistaken with the assumption the return should be different please feel free to correct me. I will gladly accept any other solution for this problem.

**Comment by goderbauer** [comment]
The various push methods on Navigator return a future that competes when the route pushed is popped again. The idea is that the pushing route can wait for a value returned by the just popped route. So this seems to be working as documented and as expected.

You are looking for a way to wait for the push animation to complete, is that correct?

**Comment by michelengelen** [comment]
@goderbauer correct ... I know that there might be a way with adding some kind of a listener/observer, but it is not something trivial to implement, afak ... so any advice would be highly appreciated.

**Comment by TahaTesser** [comment]
Hi @michelengelen 
Are you still experiencing this issue on latest stable?
If so, can you please provide your `flutter doctor -v` and `flutter run --verbose`?
Thank you

**Comment by maxlapides** [comment]
I am currently experiencing this as well.

```
[✓] Flutter (Channel beta, v1.17.0, on Mac OS X 10.15.4 19E287, locale en-US)
    • Flutter version 1.17.0 at /Users/maxlapides/flutter
    • Framework revision d3ed9ec945 (10 days ago), 2020-04-06 14:07:34 -0700
    • Engine revision c9506cb8e9
    • Dart version 2.8.0 (build 2.8.0-dev.18.0 eea9717938)


[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/maxlapides/Library/Android/sdk
    • Platform android-29, build-tools 28.0.3
    • ANDROID_HOME = /Users/maxlapides/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4.1, Build version 11E503a
    • CocoaPods version 1.9.1

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 43.0.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] VS Code (version 1.44.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.9.1

[!] Connected device
    ! No devices available

! Doctor found issues in 1 category.
```

**Comment by TahaTesser** [comment]
@maxlapides 
if you're still experiencing this problem, please create a new issue and provide the information
required in the template.

**Comment by michelengelen** [comment]
Hi @TahaTesser ... I am sry. The update mails just happened to slip by. I will try and see if it still happens when I get back to work on monday.

**Comment by iapicca** [comment]
Without additional information we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.
